crates/core/src/client/messages.rs

```diff
@@ -28,265 +24,201 @@ pub trait ToProtocol {
     fn to_protocol(self, protocol: Protocol) -> Self::Encoded;
 }
 
+pub(super) type SwitchedServerMessage = FormatSwitch<ws::ServerMessage<BsatnFormat>, ws::ServerMessage<JsonFormat>>;
+
 /// Serialize `msg` into a [`DataMessage`] containing a [`ws::ServerMessage`].
 ///
 /// If `protocol` is [`Protocol::Binary`], the message will be compressed by this method.
-pub fn serialize(msg: impl ToProtocol<Encoded = ws::ServerMessage>, protocol: Protocol) -> DataMessage {
-    let msg = msg.to_protocol(protocol);
-    match protocol {
-        Protocol::Text => serde_json::to_string(&SerializeWrapper::new(msg)).unwrap().into(),
-        Protocol::Binary => {
-            let msg_bytes = bsatn::to_vec(&msg).unwrap();
-            let reader = &mut &msg_bytes[..];
-
-            // TODO(perf): Compression should depend on message size and type.
-            //
-            // SubscriptionUpdate messages will typically be quite large,
-            // while TransactionUpdate messages will typically be quite small.
-            //
-            // If we are optimizing for SubscriptionUpdates,
-            // we want a large buffer.
-            // But if we are optimizing for TransactionUpdates,
-            // we probably want to skip compression altogether.
-            //
-            // For now we choose a reasonable middle ground,
-            // which is to compress everything using a 32KB buffer.
-            const BUFFER_SIZE: usize = 32 * 1024;
-            // Again we are optimizing for compression speed,
-            // so we choose the lowest (fastest) level of compression.
-            // Experiments on internal workloads have shown compression ratios between 7:1 and 10:1
-            // for large `SubscriptionUpdate` messages at this level.
-            const COMPRESSION_LEVEL: u32 = 1;
-            // The default value for an internal compression parameter.
-            // See `BrotliEncoderParams` for more details.
-            const LG_WIN: u32 = 22;
-
-            let mut encoder = CompressorReader::new(reader, BUFFER_SIZE, COMPRESSION_LEVEL, LG_WIN);
-
-            let mut out = Vec::new();
-            encoder
-                .read_to_end(&mut out)
-                .expect("Failed to Brotli compress `SubscriptionUpdateMessage`");
-            out.into()
+pub fn serialize(msg: impl ToProtocol<Encoded = SwitchedServerMessage>, protocol: Protocol) -> DataMessage {
+    // TODO(centril, perf): here we are allocating buffers only to throw them away eventually.
+    // Consider pooling these allocations so that we reuse them.
```

Comment:
Would so love to, but `tungstenite::Message` takes a `Vec<u8>`/`String`, so we'd have to use a different websocket library. good to have this here, though.